File: sdk/typescript/src/anchor.ts
import { encodeFunctionData, keccak256, type Hex } from "viem";

import { pqAnchorRegistryAbi, type RecordAnchorParams } from "./contracts.js";

/**
 * Post-quantum anchoring helpers.
 *
 * The PQAnchorRegistry stores a hybrid audit anchor: a Merkle root of an
 * on-chain event range PLUS a commitment to a post-quantum (ML-DSA-65,
 * NIST FIPS 204) signature over that root. The contract only commits the
 * **hash** of the PQ signature on-chain — the multi-kilobyte signature blob
 * lives off-chain (MinIO bucket `hara-pq-anchors`) and is verified off-chain
 * against the published PQ public key.
 *
 * ML-DSA signing is NOT done here — the caller signs with their PQ private key,
 * then passes the resulting signature bytes to `pqSignatureHashFromSig` (or
 * supplies a precomputed hash). For ML-DSA in JS/TS, use the optional dependency
 * `@noble/post-quantum` (ml_dsa65):
 *
 *   import { ml_dsa65 } from "@noble/post-quantum/ml-dsa";
 *   const sig = ml_dsa65.sign(secretKey, messageBytes);   // Uint8Array (~3 KB)
 *   const pqSignatureHash = pqSignatureHashFromSig(sig);  // bytes32 commitment
 *
 * pqSignatureHash MUST be non-zero; PQAnchorRegistry reverts MissingPQCommitment()
 * on a zero hash (use the legacy AnchorRegistry if you don't want PQ).
 */

/** keccak256 of the raw ML-DSA signature bytes → the on-chain bytes32 commitment. */
export function pqSignatureHashFromSig(signature: Uint8Array | Hex): Hex {
  return keccak256(signature);
}

/**
 * Build the calldata for PQAnchorRegistry.recordAnchor(...).
 * Submit with HaraChainClient.sendLegacyTx({ to: ADDRESSES.PQAnchorRegistry, data }),
 * or use the PQAnchorRegistry wrapper in contracts.ts.
 */
export function buildRecordAnchorCalldata(params: RecordAnchorParams): Hex {
  if (params.pqSignatureHash === "0x" + "0".repeat(64)) {
    throw new Error(
      "pqSignatureHash is zero — PQAnchorRegistry.recordAnchor reverts MissingPQCommitment(). " +
        "Sign the Merkle root with ML-DSA-65 and pass keccak256(signature).",
    );
  }
  if (params.blockTo < params.blockFrom) {
    throw new Error("blockTo < blockFrom — recordAnchor reverts EmptyRange().");
  }

  return encodeFunctionData({
    abi: pqAnchorRegistryAbi,
    functionName: "recordAnchor",
    args: [
      params.merkleRoot,
      params.sha3Root,
      params.blockFrom,
      params.blockTo,
      params.eventCount,
      params.anchorChain,
      params.pqSignatureHash,
    ],
  });
}

File: sdk/go/anchor.go
package hararegistry

import (
	"github.com/ethereum/go-ethereum/crypto"
)

// AnchorRecord is the input to PQAnchorRegistry.recordAnchor. All hashes are
// 32-byte values; the block range and counters are uint64.
//
// Post-quantum note: PQSignatureHash is keccak256(ML-DSA-65 signature bytes).
// The ML-DSA (Dilithium-3, NIST FIPS 204) signing happens OFF-CHAIN and is the
// caller's responsibility — this SDK only commits the hash on-chain and never
// performs PQ signing. The full signature blob is stored off-chain (MinIO
// bucket hara-pq-anchors) keyed by this hash. PQSignatureHash MUST be non-zero
// or the contract reverts MissingPQCommitment.
type AnchorRecord struct {
	// MerkleRoot is keccak256 of the event range's Merkle root.
	MerkleRoot [32]byte
	// Sha3Root is SHA3-256 of the same data (hash agility / belt-and-suspenders).
	Sha3Root [32]byte
	// BlockFrom is the first block in the anchored range (inclusive).
	BlockFrom uint64
	// BlockTo is the last block in the anchored range (inclusive). Must be >= BlockFrom.
	BlockTo uint64
	// EventCount is the number of events covered by the anchor.
	EventCount uint64
	// AnchorChain identifies the external anchoring chain (e.g. keccak/tag for
	// IOTA, Ethereum L1, ...). May be zero if not anchored externally yet.
	AnchorChain [32]byte
	// PQSignatureHash is keccak256(ML-DSA signature bytes). MUST be non-zero.
	PQSignatureHash [32]byte
}

// PackRecordAnchor ABI-encodes a recordAnchor call from an AnchorRecord. This is
// the calldata helper for callers that manage signing/nonce themselves.
func PackRecordAnchor(a AnchorRecord) ([]byte, error) {
	return abiPQAnchorRegistry.Pack(
		"recordAnchor",
		a.MerkleRoot,
		a.Sha3Root,
		a.BlockFrom,
		a.BlockTo,
		a.EventCount,
		a.AnchorChain,
		a.PQSignatureHash,
	)
}

// PQSignatureHash computes keccak256(signature) — the on-chain PQ commitment for
// a raw ML-DSA signature blob. Provided as a convenience; the ML-DSA signing
// itself is out of scope for this SDK.
func PQSignatureHash(mlDSASignature []byte) [32]byte {
	return crypto.Keccak256Hash(mlDSASignature)
}
